clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/producer/ProducerCompressionTest.java

```diff
@@ -0,0 +1,185 @@
+                        message.getBytes())));
+                // 2. send message with key, without header
+                responses.add(producer.send(new ProducerRecord<>(compressionTopic, null, now,
+                        String.valueOf(message.length()).getBytes(), message.getBytes())));
+                // 3. send message with key and header
+                responses.add(producer.send(new ProducerRecord<>(compressionTopic, null, now,
+                        String.valueOf(message.length()).getBytes(), message.getBytes(), headers)));
+            });
+            for (int offset = 0; offset < responses.size(); offset++) {
+                assertEquals(offset, responses.get(offset).get().offset(), compression);
+            }
+            verifyConsumerRecords(consumer, messages, now, headerArr, partition, compressionTopic);
+        } finally {
+            consumer.close(Duration.ofSeconds(1));
+        }
+    }
+
+    private void verifyConsumerRecords(Consumer<byte[], byte[]> consumer, List<String> messages, long now,
+                                       Header[] headerArr, int partition, String topic) {
+        TopicPartition tp = new TopicPartition(topic, partition);
+        consumer.assign(Collections.singleton(tp));
+        consumer.seek(tp, 0);
+        AtomicInteger num = new AtomicInteger(0);
+        AtomicInteger flag = new AtomicInteger(0);
+        consumeRecords(consumer, numRecords * 3, TestUtils.DEFAULT_MAX_WAIT_MS).foreach(record -> {
+            String messageValue = messages.get(num.get());
+            long offset = num.get() * 3L + flag.get();
+            if (flag.get() == 0) {
+                //  verify message without key and header
+                assertNull(record.key());
+                assertEquals(messageValue, new String(record.value()));
+                assertEquals(0, record.headers().toArray().length);
+                assertEquals(now, record.timestamp());
+                assertEquals(offset, record.offset());
+            } else if (flag.get() == 1) {
+                //  verify message with key, without header
+                assertEquals(String.valueOf(messageValue.length()), new String(record.key()));
+                assertEquals(messageValue, new String(record.value()));
+                assertEquals(0, record.headers().toArray().length);
+                assertEquals(now, record.timestamp());
+                assertEquals(offset, record.offset());
+            } else if (flag.get() == 2) {
+                //  verify message with key and header
+                assertEquals(String.valueOf(messageValue.length()), new String(record.key()));
+                assertEquals(messageValue, new String(record.value()));
+                assertEquals(1, record.headers().toArray().length);
+                assertEquals(headerArr[0], record.headers().toArray()[0]);
+                assertEquals(now, record.timestamp());
+                assertEquals(offset, record.offset());
```

Comment:
We test all compression in a single case. Could you add message to all assertion, so we can know which compression type fail?